src/dataflow/src/render/sources.rs

```diff
@@ -824,6 +893,66 @@ where
     sealed_stream
 }
 
+/// A handle that allows setting a compaction frontier.
+pub trait CompactionHandle {
+    /// Unblocks compaction for updates at or before `since`.
+    fn allow_compaction(&self, since: Antichain<Timestamp>);
+}
+
+/// A handle that allows setting a compaction frontier that is bounded by an upper bound.
+struct BoundedCompactionHandle {
+    /// Returns the upper bound when asked to.
+    upper_bound_fn: Box<dyn Fn() -> Antichain<Timestamp>>,
+
+    /// Underlying compaction handle.
+    compaction_handle: AtomicCompactionHandle,
+}
+
+impl BoundedCompactionHandle {
+    /// Creates a new [`BoundedCompactionHandle`] that bounds compaction using the given
+    /// `upper_bound_fn` and forwards compaction requests to the given `compaction_handle`
+    pub fn new<F>(upper_bound_fn: F, compaction_handle: AtomicCompactionHandle) -> Self
```

Comment:
I feel like we only need to do this upper bound stuff whilst the compaction times sent by `AllowCompaction` commands are in advance of the seal frontier right? I feel like that will go away after #10141 / something similar changes the upper reporting to only report uppers in advance of the seal frontier? 

I guess maybe a different way to say this - I'm not sure I see the need to split this out as a separate abstraction just yet, given that a lot of whats motivating the current design seems temporary. The upper bounding logic could just be in `dataflow` and then get ripped out in a few weeks?

Reply:
I had two motivations for introducing this:

 - I wanted to keep the code in `dataflow/server.rs` as minimal as possible, and I think just a call to our `allow_compaction()`, that hides the complexity inside here solves that
 - I wanted to "enshrine" the requirement we have of the seal frontier where we render the rest of the persistence code. This way, we absolutely control up to what frontier we compact, and the code still works if `coord` happens to decide that it's fine to compact to some frontier that is further ahead. So I would like to keep this even when #10141 is resolved.

Reply:
i swear I wrote this the first time and github just straight up lost it...

so I actually swing the opposite way from ruchir here and would like to see even more of this pushed into an abstraction. in particular, I think the model to follow is `ComputeCommand::AllowIndexCompaction`

concretely, we push the `Weak<Box<dyn` stuff into the abstraction, leaving us with a (strawman names for all the following, don't read into them too much) `persisted_sources: HashMap<GlobalId, PersistedSource>` map on `StorageState` like the `traces: HashMap<GlobalId, TraceBundle>` that's on `ComputeState`. we could even push this map inside a `PersistedSourceManager` to mirror `TraceManager`, which would let us move the "remove the thing from the map when we can't upgrade the weak ref" bit inside the abstraction. this leaves the dataflow command processing side as something like

```
            StorageCommand::AllowSourceCompaction(list) => {
                for (id, frontier) in list {
                    self.storage_state
                        .persisted_sources
                        .allow_compaction(id, frontier.borrow());
                }
            }
```

then on the other end, I think we push all the construction logic that's currently in seal_and_await into the abstraction too, so that part becomes something like

```
    let mut probe: ProbeHandle<Timestamp> = ProbeHandle::new();
    sealed_stream.probe_with(&mut probe);
    storage_state
        .persisted_sources
        .insert(source_id.clone(), probe);
```

thoughts?